### PR TITLE
[Feature](parquet-reader) Implements dict filter functionality parquet reader.

### PR DESCRIPTION
**Pull request by kaka11chen**
# Proposed changes

Implements dict filter functionality parquet reader to improve performance.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kaka11chen** [comment]
run buildall

File: be/src/vec/data_types/data_type.cpp [hunk 1]
```diff
@@ -56,6 +56,7 @@ void IDataType::update_avg_value_size_hint(const IColumn& column, double& avg_va
 
 ColumnPtr IDataType::create_column_const(size_t size, const Field& field) const {
     auto column = create_column();
+    column->reserve(size);
```
Comment (HappenLee) [inline be/src/vec/data_types/data_type.cpp:59]: const column nested column only need size == 1, do not need reserve `size`

File: be/src/vec/columns/column_dictionary.h [hunk 2]
```diff
@@ -166,21 +176,123 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
...
-    [[noreturn]] size_t filter(const IColumn::Filter&) override {
-        LOG(FATAL) << "filter not supported in ColumnDictionary";
+    size_t filter(const IColumn::Filter& filter) override {
+        size_t size = _codes.size();
+        if (size != filter.size()) {
+            LOG(FATAL) << "Size of filter doesn't match size of column. data size: " << size
+                       << ", filter size: " << filter.size() << get_stack_trace();
+        }
+
+        const UInt8* filter_pos = filter.data();
```
Comment (HappenLee) [inline be/src/vec/columns/column_dictionary.h:250]: filter return columnptr and size_t is similar， better to refactor

File: be/src/vec/data_types/data_type.cpp (hunk 1 above)
Comment (kaka11chen) [reply be/src/vec/data_types/data_type.cpp:59]: Use `reserve(1)` instead.

File: be/src/vec/exec/scan/vfile_scanner.cpp [hunk 3]
```diff
@@ -816,6 +876,11 @@ Status VFileScanner::_init_expr_ctxes() {
         _full_base_schema_view->table_name = _output_tuple_desc->table_desc()->name();
         _full_base_schema_view->table_id = _output_tuple_desc->table_desc()->table_id();
     }
+
+    // TODO: It should can move to scan node to process.
+    if (_vconjunct_ctx && _vconjunct_ctx->root()) {
+        _split_conjuncts(_vconjunct_ctx->root());
```
Comment (HappenLee) [inline be/src/vec/exec/scan/vfile_scanner.cpp:882]: which case `_vconjunct_ctx->root()` will be nullptr？

Comment (HappenLee) [inline be/src/vec/exec/scan/vfile_scanner.cpp:882]: which case `_vconjunct_ctx->root()` will be nullptr？

File: be/src/vec/exec/format/parquet/fix_length_dict_decoder.hpp [hunk 4]
```diff
@@ -434,13 +445,46 @@ class FixLengthDictDecoder<char*> final : public BaseDictDecoder {
...
+        }
+        doris_column->insert_many_strings(&dict_values[0], dict_items_size);
+        return Status::OK();
+    }
+
+    Status get_dict_codes(const ColumnString* string_column,
+                          std::vector<int32_t>* dict_codes) override {
+        for (int i = 0; i < string_column->size(); ++i) {
+            StringRef dict_value = string_column->get_data_at(i);
+            dict_codes->emplace_back(_dict_value_to_code[dict_value]);
```
Comment (HappenLee) [inline be/src/vec/exec/format/parquet/fix_length_dict_decoder.hpp:471]: check `dict_codes` reverse before call `get_dict_codes`?

File: be/src/vec/exec/format/parquet/vparquet_column_chunk_reader.h [hunk 5]
```diff
@@ -142,8 +143,23 @@ class ColumnChunkReader {
         return _statistics;
     }
 
+    Status read_dict_values_to_column(MutableColumnPtr& doris_column) {
+        return _decoders[static_cast<int>(tparquet::Encoding::RLE_DICTIONARY)]
+                ->read_dict_values_to_column(doris_column);
+    }
+
+    Status get_dict_codes(const ColumnString* columnString, std::vector<int32_t>* dict_codes) {
```
Comment (HappenLee) [inline be/src/vec/exec/format/parquet/vparquet_column_chunk_reader.h:151]: `columnString` to `column_string`

File: be/src/vec/exec/format/parquet/vparquet_column_reader.cpp [hunk 6]
```diff
@@ -384,6 +384,40 @@ Status ScalarColumnReader::_read_nested_column(ColumnPtr& doris_column, DataType
...
+    RETURN_IF_ERROR(_try_load_dict_page(&loaded, has_dict));
+    if (loaded && has_dict) {
+        return _chunk_reader->read_dict_values_to_column(doris_column);
+    }
+    return Status::OK();
+}
+
+Status ScalarColumnReader::get_dict_codes(const ColumnString* columnString,
+                                          std::vector<int32_t>* dict_codes) {
+    return _chunk_reader->get_dict_codes(columnString, dict_codes);
```
Comment (HappenLee) [inline be/src/vec/exec/format/parquet/vparquet_column_reader.cpp:399]: column_string

File: be/src/vec/exec/format/parquet/vparquet_group_reader.cpp [hunk 7]
```diff
@@ -71,11 +97,136 @@ Status RowGroupReader::init(const FieldDescriptor& schema, std::vector<RowRange>
...
+        return false;
+    }
+
+    if (!is_dictionary_encoded(column_metadata)) {
+        return false;
+    }
+
+    for (VExprContext* ctx : _slot_id_to_filter_conjuncts->at(slot_id)) {
+        const VExpr* root_expr = ctx->root();
+        if (root_expr->node_type() == TExprNodeType::FUNCTION_CALL) {
```
Comment (HappenLee) [inline be/src/vec/exec/format/parquet/vparquet_group_reader.cpp:148]: if the conjunt is `a > 10 is null ` ? the logic seems do the consider the case

File: be/src/vec/exec/format/parquet/vparquet_group_reader.cpp [hunk 8]
```diff
@@ -599,6 +668,236 @@ Status RowGroupReader::_filter_block_internal(Block* block,
...
+            const ColumnPtr& nested_column = nullable_column->get_nested_column_ptr();
+            const ColumnDictI32* dict_column =
+                    assert_cast<const ColumnDictI32*>(nested_column.get());
+            DCHECK(dict_column);
+
+            MutableColumnPtr string_column =
+                    _column_readers[dict_filter_col_name]->convert_dict_column_to_string_column(
+                            dict_column);
+
+            size_t pos = block->get_position_by_name(dict_filter_col_name);
```
Comment (HappenLee) [inline be/src/vec/exec/format/parquet/vparquet_group_reader.cpp:880]: 869 line search hash map by name， maybe better call `size_t pos = block->get_position_by_name(dict_filter_col_name);` in  869, only search hash map once time

File: be/src/vec/exec/format/parquet/vparquet_column_reader.cpp (hunk 6 above)
Comment (kaka11chen) [reply be/src/vec/exec/format/parquet/vparquet_column_reader.cpp:400]: ok

File: be/src/vec/exec/format/parquet/vparquet_column_chunk_reader.h (hunk 5 above)
Comment (kaka11chen) [reply be/src/vec/exec/format/parquet/vparquet_column_chunk_reader.h:151]: ok

File: be/src/vec/exec/format/parquet/fix_length_dict_decoder.hpp (hunk 4 above)
Comment (kaka11chen) [reply be/src/vec/exec/format/parquet/fix_length_dict_decoder.hpp:471]: ok

File: be/src/vec/columns/column_dictionary.h (hunk 2 above)
Comment (kaka11chen) [reply be/src/vec/columns/column_dictionary.h]: will rollback column_dictionary, so this case does not exists.

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**